rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,234 @@
+
+## Structure
+
+The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${shard}/${name}`.
+Each unit directory must contain at least a `pkg-fun.nix` file, but may contain arbitrary other files and directories.
+
+This ensures that maintainers don't have to verify this structure manually, which is prone to mistakes.
+
+## Only derivations
+
+If `pkgs/unit/${shard}/${name}` exists, `pkgs.${name}` must be a derivation that can be built directly with `nix-build`.
+
+This ensures that people can expect the unit directories to correspond to buildable packages and not functions like `pkgs.fetchFromGitHub` or `pkgs.buildRustCrate`.
+
+## Stable boundary
+
+Unit directories may only interact with the rest of Nixpkgs via the stable `pkgs.${name}` attributes, not with file references:
+- Files inside a unit directory must not reference files outside that unit directory.
+  Therefore all dependencies on other packages must come from `pkg-fun.nix` arguments injected by `callPackage`.
+  This ensures that files in Nixpkgs can be moved around without breaking this package.
+- Files outside a unit directory must not reference files inside that unit directory.
+  Therefore other packages can only depend on this package via `pkgs.${name}`.
+  This ensures that files within unit directories (except `pkg-fun.nix`) can be freely moved and changed without breaking any other packages.
+
+The only notable exception to this rule is the `pkgs/top-level/all-packages.nix` file which may reference the `pkg-fun.nix` file according to the next requirement.
+
+## Custom arguments
+
+If `pkgs/top-level/all-packages.nix` contains a definition for the attribute `${name}` and the unit directory `pkgs/unit/${shard}/${name}` exists, then the attribute value must be defined as `pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix args`, where `args` may be a freely chosen expression.
+
+This ensures that even if a package initially doesn't require a custom `args`, if it later does, it doesn't have to be moved out of the `pkgs/unit` directory to pass custom arguments.
+
+## Examples
+[examples]: #examples
+
+To add a new package `pkgs.foobar` to Nixpkgs, one only needs to create the file `pkgs/unit/fo/foobar/pkg-fun.nix`.
+No need to find an appropriate category nor to modify `pkgs/top-level/all-packages.nix` anymore.
+
+With many packages, the `pkgs/unit` directory may look like this:
+
+```
+pkgs
+└── unit
+   ├── _0
+   │  ├── _0verkill
+   │  └── _0x
+   ┊
+   ├── ch
+   │  ├── ChowPhaser
+   │  ├── CHOWTapeModel
+   │  ├── chroma
+   │  ┊
+   ┊
+   ├── t
+   │  └── t
+   ┊
+```
+
+# Interactions
+[interactions]: #interactions
+
+- `nix edit` and search.nixos.org are unaffected, since they rely on `meta.position` to get the file to edit, which still works
+- `git blame` locally and on GitHub is unaffected, since it follows file renames properly.
```

Comment:
```suggestion
- `git blame` locally and on GitHub is unaffected, since it follows file renames properly.
- Existing PR's that change migrated files won't have to be updated for the same reason
- For the same reason, changes to the migrated files can be backported to the previous release without problems
```